### PR TITLE
Min and max seq len for Pytorch

### PR DESCRIPTION
**Pull request by Atticus1806**
This PR adds `min_seq_length` and `max_seq_length` as parameters to the torch configs, similar to tensorflow.

### CODE REVIEW

**Comment by Atticus1806** [comment]
Which abstract class is he missing? `getitem`? I dont think this makes sense here tbh.
Would you raise an error or just ignore the warning? Both are done above so not sure which one is better

**Comment by albertz** [comment]
> Which abstract class is he missing? `getitem`? I dont think this makes sense here tbh. Would you raise an error or just ignore the warning? Both are done above so not sure which one is better

I assume you refer to the PyCharm warning?

> `returnn/torch/data/pipeline.py:245: WEAK WARNING PyAbstractClassInspection: Class LenFilterDataPipe must implement all abstract methods`

I don't know without looking at the base class carefully. PyCharm can tell you that automatically. Just rightclick -> implement abstract methods (or sth like that).

File: tests/test_torch_engine.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 import _setup_test_env  # noqa
 import torch
+
+import returnn.datasets.generating
```
Comment (albertz) [inline tests/test_torch_engine.py:8]: This import seems unnecessary.

File: returnn/torch/engine.py [hunk 2]
```diff
@@ -416,7 +423,9 @@ def _create_data_loader(self, dataset: Dataset) -> DataLoader2:
         )
 
         wrapped_dataset = returnn_dataset_wrapper.ReturnnDatasetIterDataPipe(dataset, reset_callback=dataset_reset)
-
+        wrapped_dataset = data_pipeline.LenFilterDataPipe(
```
Comment (albertz) [inline returnn/torch/engine.py:426]: For simplicity of the data pipeline, I would only apply this if `_min_seq_len` or `_max_seq_len` is set.

File: returnn/torch/engine.py [hunk 3]
```diff
@@ -57,6 +58,12 @@ def __init__(self, config: Config):
 
         self._start_epoch = None  # type: Optional[int]
         self._final_epoch = None  # type: Optional[int]
+        self._min_seq_len = config.typed_value("min_seq_len", None) or config.int(
+            "min_seq_len", 0
+        )  # type: Union[int,float,Dict[str,int],NumbersDict]
+        self._max_seq_len = config.typed_value("max_seq_len", None) or config.int(
+            "max_seq_len", sys.maxsize
+        )  # type: Union[int,float,Dict[str,int],NumbersDict]
```
Comment (albertz) [inline returnn/torch/engine.py:66]: The options are called `min_seq_length` and `max_seq_length`. Why this inconsistency? I would just use the same name.

File: returnn/torch/engine.py [hunk 4]
```diff
@@ -57,6 +58,12 @@ def __init__(self, config: Config):
 
         self._start_epoch = None  # type: Optional[int]
         self._final_epoch = None  # type: Optional[int]
+        self._min_seq_len = config.typed_value("min_seq_len", None) or config.int(
+            "min_seq_len", 0
+        )  # type: Union[int,float,Dict[str,int],NumbersDict]
+        self._max_seq_len = config.typed_value("max_seq_len", None) or config.int(
+            "max_seq_len", sys.maxsize
```
Comment (albertz) [inline returnn/torch/engine.py:65]: I think I would use 0 or None as the default, not `sys.maxsize`. That makes it easier to check whether it is enabled, easier in debugging.

**Comment by Atticus1806** [comment]
I think the only error left is not caused by this PR.

File: returnn/torch/data/pipeline.py [hunk 5]
```diff
@@ -26,6 +26,7 @@
 import numpy
 import torch
 import torch.utils.data
+from torch.utils.data.dataset import T_co
```
Comment (albertz) [inline returnn/torch/data/pipeline.py:29]: You don't really need that. That also causes a warning. It is not supposed to be imported.

File: returnn/torch/data/pipeline.py [hunk 6]
```diff
@@ -240,3 +241,46 @@ def __iter__(self):
...
             yield current_batch
+
+
+class LenFilterDataPipe(torch.utils.data.IterDataPipe):
+    """
+    Removes sequences which are either too long or too short from a dataset
+    Returns dataset yielding list of data lengths within the defined range
+    """
+
+    def __getitem__(self, index) -> T_co:
```
Comment (albertz) [inline returnn/torch/data/pipeline.py:252]:
```suggestion
    def __getitem__(self, index):
```

File: returnn/torch/data/pipeline.py (hunk 5 above)
Comment (albertz) [inline returnn/torch/data/pipeline.py:29]:
```suggestion
```

File: returnn/torch/data/pipeline.py [hunk 7]
```diff
@@ -240,3 +240,46 @@ def __iter__(self):
...
+
+
+class LenFilterDataPipe(torch.utils.data.IterDataPipe):
+    """
+    Removes sequences which are either too long or too short from a dataset
+    Returns dataset yielding list of data lengths within the defined range
+    """
+
+    def __getitem__(self, index):
+        raise Exception(f"{self.__class__.__name__}.__getitem__ not supported")
```
Comment (albertz) [inline returnn/torch/data/pipeline.py:252]: The `__init__` function should come first.

```suggestion
```

File: returnn/torch/data/pipeline.py [hunk 8]
```diff
@@ -240,3 +240,46 @@ def __iter__(self):
...
+        for data_dict in self._dataset:
+            # TODO: This assumes all data has time as first dimension. Currently we can't know better..
+            sequence_lengths = NumbersDict(
+                {data_key: data.shape[0] for data_key, data in data_dict.items() if data.shape}
+            )
+            if sequence_lengths.any_compare(self._min_seq_length, lambda a, b: a < b):
+                continue
+            if sequence_lengths.any_compare(self._max_seq_length, lambda a, b: a > b):
+                continue
+            yield data_dict
```
Comment (albertz) [inline returnn/torch/data/pipeline.py:285]:
```suggestion
            yield data_dict

    def __getitem__(self, index):
        raise Exception(f"{self.__class__.__name__}.__getitem__ not supported")      
```

File: returnn/torch/data/pipeline.py [hunk 9]
```diff
@@ -240,3 +240,47 @@ def __iter__(self):
...
         if current_batch:
             yield current_batch
+
+
+class LenFilterDataPipe(torch.utils.data.IterDataPipe):
+    """
+    Removes sequences which are either too long or too short from a dataset
+    Returns dataset yielding list of data lengths within the defined range
+    """
+
```
Comment (albertz) [inline returnn/torch/data/pipeline.py:250]:
```suggestion
```

File: returnn/torch/data/pipeline.py (hunk 5 above)
Comment (Atticus1806) [reply returnn/torch/data/pipeline.py]: Oh. I think this was automatically added by the PyCharm auto generation. But agreed, this can be dropped.

**Comment by albertz** [comment]
Can you apply Black?

**Review by albertz — APPROVED** [review]

**Pull request merged**